src/function/gds/single_shortest_paths.cpp

```diff
@@ -127,74 +56,52 @@ class SingleSPOutputWriterLengths : public SPOutputWriterDsts {
-        writer.beginWritingNewPath(length);
-        common::nodeID_t nodeID = dstNodeID;
-        for (auto i = 0; i < (int64_t)length - 1; ++i) {
-            nodeID = rjOutputs->ptrCast<SingleSPOutputsPaths>()->singlePaths->getParent(nodeID);
-            writer.addNode(nodeID, length - 2 - i);
-        }
-        fTable.append(vectors);
-    }
-
-private:
-    std::unique_ptr<common::ValueVector> pathNodeIDsVector;
-};
-
-struct SingleSPLengthsEdgeCompute : public EdgeCompute {
-    SinglePathLengthsFrontierPair* pathLengthsFrontiers;
-    explicit SingleSPLengthsEdgeCompute(SinglePathLengthsFrontierPair* pathLengthsFrontiers)
-        : pathLengthsFrontiers{pathLengthsFrontiers} {};
+    explicit SingleSPLengthsEdgeCompute(SinglePathLengthsFrontierPair* frontierPair)
+        : frontierPair{frontierPair} {};
 
     bool edgeCompute(common::nodeID_t, common::nodeID_t nbrID, relID_t) override {
-        return pathLengthsFrontiers->pathLengths->getMaskValueFromNextFrontierFixedMask(
-                   nbrID.offset) == PathLengths::UNVISITED;
+        return frontierPair->pathLengths->getMaskValueFromNextFrontierFixedMask(nbrID.offset) ==
+               PathLengths::UNVISITED;
     }
 
     std::unique_ptr<EdgeCompute> copy() override {
-        return std::make_unique<SingleSPLengthsEdgeCompute>(this->pathLengthsFrontiers);
+        return std::make_unique<SingleSPLengthsEdgeCompute>(frontierPair);
     }
-};
-
-struct SingleSPPathsEdgeCompute : public SingleSPLengthsEdgeCompute {
-    SinglePaths* singlePaths;
 
-    SingleSPPathsEdgeCompute(SinglePathLengthsFrontierPair* pathLengthsFrontiers,
-        SinglePaths* singlePaths)
-        : SingleSPLengthsEdgeCompute(pathLengthsFrontiers), singlePaths{singlePaths} {};
+private:
+    SinglePathLengthsFrontierPair* frontierPair;
+};
 
-    bool edgeCompute(nodeID_t curNodeID, nodeID_t nbrID, relID_t) override {
-        auto retVal = pathLengthsFrontiers->pathLengths->getMaskValueFromNextFrontierFixedMask(
-                          nbrID.offset) == PathLengths::UNVISITED;
-        if (retVal) {
-            // We set the nbrID's parent edge to curNodeID;
-            singlePaths->setParentEdge(nbrID, curNodeID);
+class SingleSPPathsEdgeCompute : public EdgeCompute {
+public:
+    SingleSPPathsEdgeCompute(SinglePathLengthsFrontierPair* frontierPair, BFSGraph* bfsGraph)
+        : frontierPair{frontierPair}, bfsGraph{bfsGraph} {
+        parentListBlock = bfsGraph->addNewBlock();
+    }
+
+    bool edgeCompute(nodeID_t boundNodeID, nodeID_t nbrNodeID, relID_t edgeID) override {
+        auto shouldUpdate = frontierPair->pathLengths->getMaskValueFromNextFrontierFixedMask(
+                                nbrNodeID.offset) == PathLengths::UNVISITED;
+        if (shouldUpdate) {
+            if (!parentListBlock->hasSpace()) {
+                parentListBlock = bfsGraph->addNewBlock();
+            }
+            bfsGraph->addParent(frontierPair->curIter.load(std::memory_order_relaxed),
```

Comment:
Hmm, you are still possibly keeping track of multiple parents per node if you dont have a special addParent function for single shortest paths. So the parent writing still happens inside a while loop and does not overwrite any previously written parents. This might be OK as long as the output writing takes care of outputting only one path (but it would still be slower). That said, I don't see that happening. So it looks like the current outputwriter used for single shortest paths, when outputting paths, can output multiple paths per node because it uses exactly the same `SPPathsOutputWriter` that AllSPAlgorithm uses. Does this make sense? I'm thinking of the case when 2 threads, T1 and T2 readh shouldUpdate as true for a neighbor node u here, then then they both call bfsGraph->addParent. That will end up inserting 2 parent edges for u. We need to fix this.  

So my suggestion was to have an if/else block inside addParent for single vs all shortest paths. All shortest paths uses a while loop while single sp does not.

Reply:
Done. Though it might be good if u take another look at `tryAddSingleParent` function.